mod-clans/src/main/java/gg/solarmc/loader/clans/Clan.java

```diff
@@ -0,0 +1,425 @@
+
+        int sec = transaction.getProperty(DSLContext.class)
+                .insertInto(CLANS_CLAN_MEMBERSHIP)
+                .columns(CLANS_CLAN_MEMBERSHIP.CLAN_ID,CLANS_CLAN_MEMBERSHIP.USER_ID)
+                .values(this.clanID,receiver.getUserId())
+                .execute();
+
+        receiver.setCachedClan(this);
+
+        return getClanMembers(transaction);
+    }
+
+    /**
+     * Removes a member from the clan.
+     * @param transaction The tx
+     * @param player player to remove from the clan
+     * @return Accurate set of ClanMembers post transaction
+     * @throws IllegalStateException if member already has a clan
+     * @throws IllegalArgumentException if member tried to remove was the leader of the clan, or if they are not
+     * present in the clan
+     */
+    public Set<ClanMember> removeClanMember(Transaction transaction, SolarPlayer player) {
+        ClanDataObject receiver = player.getData(ClansKey.INSTANCE);
+        if (receiver.getClan(transaction).isPresent()) throw new IllegalStateException("Receiver already has clan!");
+
+        ClansClanMembershipRecord rec = transaction.getProperty(DSLContext.class)
+                .fetchOne(CLANS_CLAN_MEMBERSHIP,CLANS_CLAN_MEMBERSHIP.CLAN_ID.eq(this.clanID).and(CLANS_CLAN_MEMBERSHIP.USER_ID.eq(receiver.getUserId())));
+
+        if (rec == null) throw new IllegalArgumentException("Member is not part of clan!");
+        if (receiver.isSimilar(leader)) throw new IllegalArgumentException("Tried to remove leader of clan!");
+
+        rec.delete();
+
+        receiver.setCachedClan(null);
+
+        return getClanMembers(transaction);
+    }
+
+    /**
+     * Adds clan as an ally
+     *
+     * @param transaction the tx
+     * @param receiver the receiver
+     * @return whether record was inserted or not.
+     * @throws IllegalArgumentException if receiver is the same gg.solarmc.loader.clans.Clan as this object
+     * @throws IllegalStateException if receiver or this object already has an ally
+     */
+    public boolean addClanAsAlly(Transaction transaction, Clan receiver) {
+        if (receiver.getID() == this.clanID) throw new IllegalArgumentException("Tried to mark clan ally as same clan!");
+        if (this.currentlyAlliedClanID().isPresent()) throw new IllegalStateException("Sender already have an ally!");
+        if (receiver.currentlyAlliedClanID().isPresent()) throw new IllegalStateException("Selected clan already has an ally!");
+
+        //note to aurium - these are now okay to do becaue we run the checks previous
+        int res = transaction.getProperty(DSLContext.class)
+                .insertInto(CLANS_CLAN_ALLIANCES)
+                .columns(CLANS_CLAN_ALLIANCES.CLAN_ID,CLANS_CLAN_ALLIANCES.ALLY_ID)
+                .values(this.clanID,receiver.getID())
+                .execute();
+
+        manager.insertAllianceCache(this.getID(),receiver.getID());
+
+        return res == 1;
```

Comment:
That is exactly what it is checking for. Do you want me to return void and instead throw an exception on constraint violation (update == 0)?

Reply:
Oh, I just realized I said the update count is 0. That's not right.

With your current approach (standard INSERT):
If there is a constraint violation, an exception is thrown and you will not get an update count. 
Otherwise, the row is inserted and the update count is 1

With an upsert (INSERT IGNORE)
If there is a constraint violation, the update count is 0
Otherwise, the row is inserted and the update count is 1

However, in this case, there will not be a constraint violation unless this clan is allied precisely to the specified clan. We may need to revisit the schema